### ISSUE TITLE
SanitiseArgs not working with JSON payload

### ISSUE THREAD

**Issue by zimmerle** [SpiderLabs/ModSecurity#715]
Reported by: Bruno Savioli de Almeida.
http://sourceforge.net/p/mod-security/mailman/message/32281341/

ModSecurity version 2.8.0 

Rule:

```
SecAction "phase:5,id:'6660666',t:none,pass,nolog,sanitiseArg:cardNumber,sanitiseArg:cardToken"
```

audit log:

```
[29/Apr/2014:12:19:54 +0100] U1@K2goFLh4AAHIFMqAAAAAS 10.5.12.18 43609 10.5.46.31 443
--72235b1e-B--
POST /psp/save HTTP/1.1
User-Agent: curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
Host: payments
Content-Type: application/json;charset=UTF-8
Accept: application/json
Content-Length: 114

--72235b1e-C--
{"cardToken":"aaaaaaaaaaaaaaaaaaaaaaaaaa1111111111111111111111aaaaaaaaaaaaaaaaaa","cardNumber":"1000000000000001"}
--72235b1e-F--
HTTP/1.1 400 Bad Request
Content-Type: application/json
Via: 1.1 payments
Content-Length: 78
Connection: close

--72235b1e-E--
{"message":"Please check your input and try again.","error":"Invalid Details"}
--72235b1e-H--
Apache-Handler: proxy-server
Stopwatch: 1398770394130647 22955 (- - -)
Stopwatch2: 1398770394130647 22955; combined=2733, p1=266, p2=2062, p3=9, p4=355, p5=40, sr=86, sw=1, l=0, gc=0
Response-Body-Transformed: Dechunked
Producer: ModSecurity for Apache/2.8.0 (http://www.modsecurity.org/); OWASP_CRS/2.2.9.
Server: Apache
Sanitised-Args: "cardNumber", "cardToken".
Engine-Mode: "DETECTION_ONLY"
```

A Similar request using  application/x-www-form-urlencoded works as expected.

**Comment by ccaau** [comment]
Hi Zimmerle, I seem to have the same issue as described above with nginx/modsec 1.8.0 .. do you know if anyone has resolved this either by a patch or by config?

thanks,
C

**Comment by zimmerle** [comment]
Hi @ccaau,

There was no patch to fix this issue yet. Currently we are working on libmodsecurity  + Nginx connector. Marking this issue to be resolved as part of "libmodsecurity"  release.

**Comment by ccaau** [comment]
Thanks @zimmerle 

Can you please confirm that there is currently no support to sanitise the REQUEST_BODY audit log section when the payload is JSON. For example, we can't use an arbitrary regex over the payload.

thanks,
C

**Comment by zimmerle** [comment]
Hi @ccaau,

I can confirm that this functionality is not really working as expected. There is a bug/missing feature there that should be investigated. Most likely it will be ready for ModSecurity version 3. The problem is that the section C of auditlog [1] is not being sanitized when the request body is a JSON.

[1] https://github.com/SpiderLabs/ModSecurity/wiki/ModSecurity-2-Data-Formats#request-body-c

**Comment by ccaau** [comment]
Hi @zimmerle,

thanks very much for the clarification! 

Just for anyone else, we're selectively not logging the C part using ctl:auditLogParts=-CI based on the URI

Regards,
C

**Comment by mohakkataria** [comment]
Hey Guys,

Was this issue resolved? Facing the same issue here.

**Comment by zimmerle** [comment]
Hi @mohakkataria, the issue is open.

**Comment by x3rus** [comment]
Hi folks, 

This issue was open a long time ago , I understand it's not an easy fix, but do you have anyway idea when it will be resolve ? 

@zimmerle you said it will be ready in version 3 , (https://github.com/SpiderLabs/owasp-modsecurity-crs/releases/tag/v3.0.0) this release include the fix ?

Thanks again for all you works , very nice job !

**Comment by zimmerle** [comment]
Hi @x3rus, the SanitiseArgs is not yet implemented in v3.

**Comment by blorenz1984** [comment]
Hey,

is there any estimation on until when sanitising of JSON will be implemented? As the initial post implies, this is also of PCI relevance...

Thanks for all you are doing here!

**Comment by phillbaker** [comment]
Hello just checking if this made it's way into v3 now that v3 has been officially released?

**Comment by zimmerle** [comment]
Indeed @phillbaker, not yet supported.

**Comment by calve** [comment]
hi, had a bit trouble to find this issue while trying to sanitize/sanatize my audit log on libmodsecurity. @zimmerle, is there any particuliar reason that the first `i` in `sanitise` mutated to an `a` (`sanatise`) at https://github.com/SpiderLabs/ModSecurity/blob/077b18252ffab347156a56e651120248cc4491ef/src/parser/seclang-scanner.ll#L119

**Comment by gagandeepsingh485** [comment]
@zimmerle : Any timelines on when will we be having support for the following actions ?
ACTION_SANATISE_ARG 
ACTION_SANATISE_MATCHED 
ACTION_SANATISE_MATCHED_BYTES
ACTION_SANATISE_REQUEST_HEADER
ACTION_SANATISE_RESPONSE_HEADER

So our major use case is to filter out certain sensitive content from the logs. Is there any alternate way to do these things in V3 or should we just go back to the older version ?

**Comment by zimmerle** [comment]
Hi @calve,

Likely to be a typo. Want to send a patch for it?

**Comment by zimmerle** [comment]
Hi @gagandeepsingh485,

Not yet. Sorry :(

**Comment by zimmerle** [comment]
> hi, had a bit trouble to find this issue while trying to sanitize/sanatize my audit log on libmodsecurity. @zimmerle, is there any particuliar reason that the first `i` in `sanitise` mutated to an `a` (`sanatise`) at
> 
> [ModSecurity/src/parser/seclang-scanner.ll](https://github.com/SpiderLabs/ModSecurity/blob/077b18252ffab347156a56e651120248cc4491ef/src/parser/seclang-scanner.ll#L119)
> 
> Line 119 in [077b182](/SpiderLabs/ModSecurity/commit/077b18252ffab347156a56e651120248cc4491ef)
> 
>  ACTION_SANATISE_ARG                             (?i:sanitiseArg)

Fixed as of: c2bc6952651707f773280b7b8188db814aa11718

**Comment by zimmerle** [comment]
There are **two different things** being handling in this issue: the original subject of the issue and the support for _SanitiseArgs_ on _v3_.

The original issue was closed as of f86de566d18dda6351ecba52d5e5f1d29ad02a12 (or: #1429). There is a new issue to track the support of _SanitiseArgs_ on _v3_ here: #1898.

Therefore, there is no need to keep this issue open.
